### PR TITLE
Add support for NSX-T ALB Virtual Services

### PR DESCRIPTION
**Pull request by Didainius**
**Note** Only to be reviewed after #756 is merged as this PR contains it.

This PR adds `vcd_nsxt_alb_virtual_service` resource and data source which complete the NSX-T ALB puzzle. It ought to be feature complete after this PR is merged.

Extra. 
* While finally integrating all components and using tenant user it appeared that  `vcd_nsxt_alb_edgegateway_service_engine_group` requires `service_engine_group_id` which is not really possible to get with tenant user. To mitigate it now this resource supports both `service_engine_group_name` and `service_engine_group_id`  (requires only 1 of them).
* This PR adds an ALB doc to "guides" section with rough overview and end to end configuration example

Acceptance (with and without Org user), binary and upgrade tests passed for tag ALB.

### CODE REVIEW

File: website/docs/r/nsxt_alb_virtual_service.html.markdown [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+  name            = "new-virtual-service"
+  edge_gateway_id = data.vcd_nsxt_edgegateway.existing.id
+
+  pool_id                  = vcd_nsxt_alb_pool.test.id
+  service_engine_group_id  = vcd_nsxt_alb_edgegateway_service_engine_group.assignment.service_engine_group_id
+  virtual_ip_address       = tolist(data.vcd_nsxt_edgegateway.existing.subnet)[0].primary_ip
+  ca_certificate_id        = data.vcd_library_certificate.org-cert-1.id
+  application_profile_type = "L4_TLS"
+  service_port {
```
Comment (lvirbalas) [inline website/docs/r/nsxt_alb_virtual_service.html.markdown:115]: I'd add a line above for added clarity:

```suggestion

  service_port {
```

File: vcd/resource_vcd_nsxt_alb_virtual_service.go [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+			"description": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Description of ALB Virtual Service",
+			},
+			"pool_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Edge gateway ID in which ALB Pool should be created",
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_alb_virtual_service.go:60]: Wrong description :)

File: vcd/resource_vcd_nsxt_alb_virtual_service.go [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+			},
+			"service_engine_group_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Edge gateway ID in which ALB Pool should be created",
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_alb_virtual_service.go:66]: Wrong description :)

File: vcd/resource_vcd_nsxt_alb_virtual_service.go [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+	Schema: map[string]*schema.Schema{
+		"start_port": &schema.Schema{
+			Type:        schema.TypeInt,
+			Required:    true,
+			Description: "Starting port in the range",
+		},
+		"end_port": &schema.Schema{
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Description: "Starting port in the range",
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_alb_virtual_service.go:108]: Wrong description.

File: vcd/resource_vcd_nsxt_alb_virtual_service.go [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+	}
+
+	edge, err := vdc.GetNsxtEdgeGatewayByName(edgeName)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve NSX-T edge gateway with ID '%s': %s", d.Id(), err)
+	}
+
+	albVirtualService, err := vcdClient.GetAlbVirtualServiceByName(edge.EdgeGateway.ID, virtualServiceName)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve NSX-T ALB Virtual Service '%s': %s", virtualServiceName, err)
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_alb_virtual_service.go:232]: There's some inconsistency in capitalization: in some places it's `Virtual Service`, while in others - `virtual service`.

**Review by lvirbalas — CHANGES_REQUESTED** [review]
Some feedback. BTW, nice that you have SSL-enabled tests!

File: website/docs/r/nsxt_alb_virtual_service.html.markdown (hunk 1 above)
Comment (Didainius) [reply website/docs/r/nsxt_alb_virtual_service.html.markdown:115]: Fixed.

File: vcd/resource_vcd_nsxt_alb_virtual_service.go (hunk 3 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_alb_virtual_service.go]: Fixed this and a few more erroneous descriptions. Hopefully it should be good for all.

File: vcd/resource_vcd_nsxt_alb_virtual_service.go (hunk 4 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_alb_virtual_service.go]: Fixed this and a few more erroneous descriptions. Hopefully it should be good for all.

File: vcd/resource_vcd_nsxt_alb_virtual_service.go (hunk 5 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_alb_virtual_service.go:232]: Yes. Used search now to hopefully catch all.

**Review by lvirbalas — APPROVED** [review]

**Review by vbauzysvmware — APPROVED** [review]

File: vcd/datasource_vcd_nsxt_alb_virtual_service.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+func datasourceVcdAlbVirtualServiceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return diag.Errorf("error getting Org and VDC: %s", err)
+	}
+
+	if vdc.IsNsxv() {
+		return diag.Errorf("ALB Pools are only supported on NSX-T. Please use 'vcd_lb_server_pool' for NSX-V load balancers")
```
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_alb_virtual_service.go]: Fixed.

File: vcd/datasource_vcd_nsxt_alb_virtual_service.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+						},
+						"end_port": &schema.Schema{
+							Type:        schema.TypeInt,
+							Computed:    true,
+							Description: "Starting port in the range",
+						},
+						"ssl_enabled": &schema.Schema{
+							Type:        schema.TypeBool,
+							Computed:    true,
+							Description: "Starting port in the range",
```
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_alb_virtual_service.go:92]: Fixed.

File: vcd/datasource_vcd_nsxt_alb_virtual_service.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+			},
+			"description": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Description of ALB Virtual Service",
+			},
+			"pool_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Edge gateway ID in which ALB Pool should be created",
```
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_alb_virtual_service.go]: Fixed.

**Review by dataclouder — APPROVED** [review]

**Pull request merged**